sycl/include/sycl/detail/cg_types.hpp

```diff
@@ -235,6 +235,66 @@ class HostKernel : public HostKernelBase {
 #endif
 };
 
+// the class keeps reference to a lambda allocated externally on stack
+class HostKernelRefBase : public HostKernelBase {
+public:
+  virtual std::unique_ptr<HostKernelBase> takeOrCopyOwnership() const = 0;
+#ifndef __INTEL_PREVIEW_BREAKING_CHANGES
+  // This function can't be called from old user code, because there is no
+  // HostKernelRef in old user code. So, make it empty.
+  void InstantiateKernelOnHost() override {}
+#endif
+};
+
+// primary template for movable lambda objects
+template <class KernelType, class KernelTypeUniversalRef, class KernelArgType,
+          int Dims>
+class HostKernelRef : public HostKernelRefBase {
```

Comment:
I'd prefer to have `HostKernelRefBase`->`HostKernelRef` (we create objects once and then use in several places by referring to the type-erased base type) and `HostKernelRef`->`HostKernelRefImpl` or something like that, but I see how aligning with a poor name choice for the existing `HostKernelBase` is a counterargument to that.

Maybe adding

```c++
template <typename KernelType>
static HostKernelRef HostKernelRef::create(KernelType &&Kernel) {}
```

and avoiding using the `HostKernelRefImpl` would help with clarity too (and we'd isolate choosing the right template (rvalue ref vs lvalue ref) to this single `create` function).

 @vinser52 , WDYT?

Reply:
Sorry, I did not get you idea and why do we need one more derived class (`HostKernelRefImpl`)? What are you trying to achieve with that?

> Maybe adding
> ```cpp
> template <typename KernelType>
> static HostKernelRef HostKernelRef::create(KernelType &&Kernel) {}
> ```

Are you trying to avoid `HostKernelref` specialization for the `const MKerne &`?

Reply:
Most APIs will be accepting `HostKernelRefBase`, I'm trying to avoid the `Base` suffix in those names. I'm not suggesting any new types in the hierarchy.

Reply:
Ahh, now I got your point. 

> `HostKernelRefBase`->`HostKernelRef`

This means rename. For me `*Base` clearly indicates that we are dealing with the base class. But it is subjective.